FF-A: SPMC test

jellesels-arm opened the pull request:
Add support for testing the SPMC component inside OP-TEE. 
A new set of test have been added under regression_9000. This test some of the SPMC functionality. To be able to test this PR the [BUILD PR](https://github.com/OP-TEE/build/pull/584) is also needed. 
The SPs which are used on the Secure Wold to test the SPMC are currently under review on the Trusted Service gerrit server: https://review.trustedfirmware.org/c/TS/trusted-services/+/10159

Test can be run by using the following manifest file.
repo init -u https://github.com/jellesels-arm/manifest.git -m fvp-spmc-test.xml -b SPMC_TEST  &&  repo sync -j8 --no-clone-bundle
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on host/xtest/include/ffa.h:
```diff
@@ -0,0 +1,175 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
Instead of copying this file into this git we could make it available in the dev-kit by adding it to the list `incfiles-extra-host` in `ta/ta.mk`.

jellesels-arm replied:
Ack, [OP-TEE OS PR](https://github.com/OP-TEE/optee_os/pull/5522)

jenswi-linaro commented on host/xtest/include/arm_ffa_user.h:
```diff
@@ -0,0 +1,79 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
This file should probably go into `host/xtest/include/uapi/linux` until it's available by other means.

jellesels-arm replied:
Ack

jenswi-linaro commented on host/xtest/CMakeLists.txt:
```diff
@@ -62,6 +62,9 @@ set (SRC
 	xtest_uuid_helpers.c
 )
+if(CFG_SECURE_PARTITION AND CFG_SPMC_TESTS)
+	list (APPEND SRC regression_9000.c)
```
I think it would make sense to have this in a separate test suite, compare with the PKCS#11 suite below. It could still be run by default if enabled as we do for the GP tests.

jellesels-arm replied:
Ack

jenswi-linaro commented on host/xtest/regression_9000.c:
```diff
@@ -0,0 +1,174 @@
+	EP_TEST_SP_COMMUNICATION,
+};
+
+static int fp = -1;
```
I'd prefer `ffa_fd`.

jellesels-arm replied:
Ack

jenswi-linaro commented on host/xtest/regression_9000.c:
```diff
@@ -0,0 +1,174 @@
+		goto out;
+	}
+
+	endpoint1_id = get_endpoint_id(test_endpoint1.uuid_ptr);
```
Perhaps `Do_ADBG_BeginSubCase(c, "SP1 comms check");` just before this and a corresponding `Do_ADBG_EndSubCase()` when complete.

Same for the other logical subtests.

jellesels-arm replied:
Ack

jenswi-linaro commented on host/xtest/regression_9000.c:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	memset(&args, 0, sizeof(args));
+	ADBG_EXPECT_TRUE(c, start_sp_test(endpoint1_id, EP_TEST_SP, &args));
```
If either of these fails, does it make any sense to continue?

jellesels-arm replied:
Ack

jforissier commented on host/xtest/regression_9000.c:
```diff
@@ -0,0 +1,174 @@
+	if (fp < 0) {
+		fp = open(FFA_DRIVER_FS_PATH, O_RDWR);
+		if (fp < 0) {
+			fprintf(stderr, "Error: Could not open the FF-A driver\n");
```
Why not `Do_ADBG_Log()`?

jellesels-arm replied:
Ack

jforissier reviewed (commented):
About "FF-A: Add SPMC test SPs": I'd rather not commit any binary into the Git source repo.  These tests depend on so many other things than what is strictly is this Git anyways.

jellesels-arm commented on host/xtest/regression_9000.c:
```diff
@@ -0,0 +1,174 @@
+		return false;
+	}
+	return true;
+}
```
Ack

jellesels-arm commented:
> 
Deleted the patch and changed the build to always build the SPs

jellesels-arm commented:
> About "FF-A: Add SPMC test SPs": I'd rather not commit any binary into the Git source repo. These tests depend on so many other things than what is strictly is this Git anyways.

 Deleted the patch and changed the build to always build the SPs

jellesels-arm commented:
Update [xtest: SPMC Add basic test](https://github.com/OP-TEE/optee_test/pull/614/commits/133f931e0af3563b0962543f02356a0962c81ba7) commit message to show that the TS debugfs driver has to be loaded before the tests can be run.

jenswi-linaro commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -137,10 +121,12 @@ static void xtest_tee_test_9001(ADBG_Case_t *c)
+	ADBG_REQUIRE_TRUE(c, exit_test, start_sp_test(endpoint1_id, EP_TEST_SP,
+			  &args))
+	ADBG_REQUIRE_TRUE(c, exit_test, check_ok(&args))
+	Do_ADBG_EndSubCase(c, NULL);
```
While a subcase can be ended with a NULL pointer, it's preferable to supply the same name that was used when the subcase was opened. It's easier to see which subcase you're ending then.

jellesels-arm replied:
Ack

jenswi-linaro commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -137,10 +121,12 @@ static void xtest_tee_test_9001(ADBG_Case_t *c)
 	memset(&args, 0, sizeof(args));
-	ADBG_EXPECT_TRUE(c, start_sp_test(endpoint1_id, EP_TEST_SP, &args));
-	ADBG_EXPECT_TRUE(c, check_ok(&args));
+	ADBG_REQUIRE_TRUE(c, exit_test, start_sp_test(endpoint1_id, EP_TEST_SP,
```
Usually, when we write tests we try to make the error reporting contain details that give some help to debug the issue. With this, you just report that it failed without any error code.

I would have written this as:
``` C
rc = start_sp_test(endpoint1_id, EP_TEST_SP, &args);
if (!ADBG_EXPECT_COMPARE_SIGNED(c, rc, ==, 0))
     goto out;
```
where start_sp_test() returned the error value instead. Sure, in some cases the error value will not be very useful, but it can never contain less information than `false`. By the way, I avoid the ADBG_REQUIRE_*() macros since they encourage helper macros like your `exit_test`.

jellesels-arm replied:
Ack

jforissier commented on host/xtest/Makefile:
```diff
@@ -83,6 +83,10 @@ srcs +=	adbg/src/adbg_case.c \
 	xtest_uuid_helpers.c
 
+ifeq ($(CFG_SECURE_PARTITION)-$(CFG_SPMC_TESTS),y-y)
+srcs +=ffa_spmc_1000.c
```
space

jellesels-arm replied:
Ack

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,164 @@
+
+static bool init_sp_xtest(ADBG_Case_t *c)
+{
+
```
Extra blank line

jellesels-arm replied:
Ack

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,164 @@
+
+static void close_debugfs(void)
+{
+	int err;
```
Add empty line after declarations

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,164 @@
+		err = close(ffa_fd);
+		if (err < 0) {
+			Do_ADBG_Log("Error: Could not close the FF-A driver\n");
+		}
```
`{ }` not needed

jellesels-arm replied:
Ack

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,164 @@
+
+static uint16_t get_endpoint_id(uint64_t endp)
+{
+	struct ffa_ioctl_ep_desc sid = {0};
+
+	sid.uuid_ptr = endp;
```
struct ffa_ioctl_ep_desc sid = { .uuid_ptr = endp };

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,164 @@
+
+struct ffa_ioctl_ep_desc test_endpoint1 = {
+		.uuid_ptr = (uint64_t) test_endpoint1_uuid,
+		.id = 0
```
Personally I wouldn't set `.id` since it is guaranteed to be zero

jellesels-arm replied:
Ack

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -160,3 +179,292 @@ static void xtest_ffa_spmc_test_1001(ADBG_Case_t *c)
+		*handle = shm_desc.handle;
+
+	return status;
+
```
Extra blank line

jellesels-arm replied:
Ack

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -160,3 +179,292 @@ static void xtest_ffa_spmc_test_1001(ADBG_Case_t *c)
+static bool share_mem(uint16_t endpoint, uint64_t *handle)
+{
+	bool status = false;
+	struct ffa_ioctl_shm_desc shm_desc = {.dst_id = endpoint,
```
Add blank line here

jellesels-arm replied:
@jforissier there is already a blank line after the struct declaration. Am I missing something?

jforissier replied:
My mistake, sorry

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -160,3 +179,292 @@ static void xtest_ffa_spmc_test_1001(ADBG_Case_t *c)
+	int ret = 0;
+	struct ffa_ioctl_shm_desc shm_desc = {0};
+
+
```
One blank line is enough

jellesels-arm replied:
Ack

jforissier commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -160,3 +179,292 @@ static void xtest_ffa_spmc_test_1001(ADBG_Case_t *c)
+			  &args))
+	ADBG_REQUIRE_TRUE(c, exit_test, check_ok(&args))
+
+
```
One blank line is enough

jellesels-arm replied:
Ack

jforissier reviewed (commented):
Hi @jellesels-arm,

I have a few cosmetic comments, in any case LGTM.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -154,3 +173,303 @@ static void xtest_ffa_spmc_test_1001(ADBG_Case_t *c)
+
+out:
+	close_debugfs();
+
```
remove empty line

jellesels-arm replied:
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -154,3 +173,303 @@ static void xtest_ffa_spmc_test_1001(ADBG_Case_t *c)
+}
+
+ADBG_CASE_DEFINE(ffa_spmc, 1002, xtest_ffa_spmc_test_1002,
+		 "Test FF-A memory: ");
```
Ack

jellesels-arm commented on host/xtest/include/uapi/linux/arm_ffa_user.h:
```diff
@@ -0,0 +1,79 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Copyright (c) 2020-2021, Arm Limited
```
Ack + removed ffa.h

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,156 @@
+
+static uint16_t get_endpoint_id(uint64_t endp)
+{
+	struct ffa_ioctl_ep_desc sid = {.uuid_ptr = endp };
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+	if (ffa_fd >= 0) {
+		err = close(ffa_fd);
+		if (err < 0)
+			Do_ADBG_Log("Error: Could not close the FF-A driver\n");
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+const char test_endpoint3_uuid[] = "23eb0100-e32a-4497-9052-2f11e584afa6";
+
+struct ffa_ioctl_ep_desc test_endpoint1 = {
+		.uuid_ptr = (uint64_t) test_endpoint1_uuid,
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+
+static void close_debugfs(void)
+{
+	int err;
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+	Do_ADBG_EndSubCase(c, "SP to SP messaging check");
+
+out:
+	close_debugfs();
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+	if (!ADBG_EXPECT_COMPARE_SIGNED(c, rc, ==, 0))
+	     goto out;
+
+	if (!ADBG_EXPECT_COMPARE_SIGNED(c, args.args[0], ==, SPMC_TEST_OK))
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+	memset(&args, 0, sizeof(args));
+	rc = start_sp_test(endpoint1_id, EP_TEST_SP, &args);
+	if (!ADBG_EXPECT_COMPARE_SIGNED(c, rc, ==, 0))
+	     goto out;
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,475 @@
+	return true;
+}
+
+static signed int start_sp_test(uint16_t endpoint, enum sp_tests test,
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,156 @@
+
+const char test_endpoint1_uuid[] = "5c9edbc3-7b3a-4367-9f83-7c191ae86a37";
+const char test_endpoint2_uuid[] = "7817164c-c40c-4d1a-867a-9bb2278cf41a";
+const char test_endpoint3_uuid[] = "23eb0100-e32a-4497-9052-2f11e584afa6";
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,156 @@
+ */
+#include <fcntl.h>
+#include "include/uapi/linux/arm_ffa_user.h"
+#include "include/ffa.h"
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+	check_alive(c, endpoint1_id);
+
+out:
+	Do_ADBG_EndSubCase(c, "Test retrieve invalid memory region");
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+	*handle = 0;
+	/* Share memory with SP*/
+	rc = share_mem(endpoint, handle);
+	ADBG_EXPECT_COMPARE_SIGNED(c, rc, ==, 0);
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+{
+	int status = false;
+	struct ffa_ioctl_shm_desc shm_desc = { .dst_id = endpoint,
+					      .size = 0x1000};
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+	rc = share_mem(endpoint, handle);
+	ADBG_EXPECT_COMPARE_SIGNED(c, rc, ==, 0);
+
+	ADBG_EXPECT_TRUE(c, handle != 0);
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+
+static void check_alive(ADBG_Case_t *c, uint16_t endpoint)
+{
+	struct ffa_ioctl_msg_args args = {0};
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+	rc = start_sp_test(endpoint2_id, EP_TEST_SP_COMMUNICATION, &args);
+	ADBG_EXPECT_COMPARE_SIGNED(c, rc, ==, 0);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, args.args[0], ==, SPMC_TEST_OK);
+
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+	"23eb0100-e32a-4497-9052-2f11e584afa6";
+
+static struct ffa_ioctl_ep_desc test_endpoint1 = {
+	.uuid_ptr = (uint64_t) test_endpoint1_uuid,
```
Ack

jellesels-arm commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,473 @@
+#define NORMAL_WORLD_ENDPOINT_ID	0
+
+#define MEM_SHARE_HANDLE_LSB(x) (x & 0xffffffff)
+#define MEM_SHARE_HANDLE_MSB(x) ((x >> 32) & 0xffffffff)
```
Ack

jenswi-linaro commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,479 @@
+#define INCORRECT_ENDPOINT_ID 0xffff
+#define NORMAL_WORLD_ENDPOINT_ID	0
+
+#define MEM_SHARE_HANDLE_LSB(x) ((x) & 0xffffffff)
```
What is "LSB" short for?

jellesels-arm replied:
Added a comment

jenswi-linaro replied:
bytes or bits? :-)
How about instead talking about low/high part of a handle? Using the abbreviations LSB and MSB may bring different expectations.

jellesels-arm replied:
Ack

jenswi-linaro commented on host/xtest/ffa_spmc_1000.c:
```diff
@@ -0,0 +1,479 @@
+			 struct ffa_ioctl_msg_args *args)
+{
+	args->dst_id = endpoint;
+	args->args[0] = (uint32_t)test;
```
Cast is not needed.

jellesels-arm replied:
Ack

jellesels-arm commented:
Any more comments on this? I would we good if we could merge the SPMC testing before the 3.19 release.

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`
